### PR TITLE
Add `SepCache` [An efficient and easy-to-use Cache from the SepLLM paper - ICML 2025 (https://arxiv.org/abs/2412.12094) ] to the `cache_utils.py` and `__init__.py`

### PR DESCRIPTION
**Pull request by GaussonTschen**
Add `SepCache` [An efficient and easy-to-use Cache from the SepLLM paper - ICML 2025 (https://arxiv.org/abs/2412.12094) ] to the `cache_utils.py` and `__init__.py`

    A cache as described in the [SepLLM paper - ICML 2025](https://arxiv.org/abs/2412.12094). In the training phase, 
    SepLLM condenses the segment information into the KV of the separator that divides the segment. In the inference phase, the 
    corresponding SepCache only needs to store the KVs of initial tokens, separator tokens, and recent tokens for generation.

    It stores the Key and Value states as lists of tensors, two lists for each layer. The expected shape for each tensor is
    `[batch_size, num_heads, seq_len, head_dim]`.

    Frequently-Used Parameters:

        `init_cache_size: Union[int, List]`:
            The maximum number of KVs to be stored for initial tokens.                
                
        `sep_cache_size: Union[int, List]`:
            The maximum number of KVs to be stored for separator tokens.

        `local_size: Union[int, List]`: 
            The maximum number of KVs to be stored for local tokens (i.e., sliding window).

        `cache_size: Union[int, List]`:    
            The maximum number of KVs to be stored for all the tokens, i.e., the size for the whole KV cache.  

        Concerning these four parameters above:
            When a list is passed (its length must be `layer_num`), it represents different values for each layer. 
            When an integer is passed, it means the setting is the same for all layers.
        
        
        `separator_token_ids: List[int]`:
            The token ids of the separator tokens for the current model's tokenizer.            
            We have some examples, such as the Llama-3 series models, where setting `model_type='llama'` allows you 
                to skip setting `separator_token_ids` and `PADDING_ID` (SepCache will auto-fill them).

        `PADDING_ID: int`:
            The token id of the padding token. You can just set `PADDING_ID` to the id of "<|endoftext|>" token of the tokenizer for the pretrained model.

    Example:

        ```python        
        >>> from transformers import AutoTokenizer, AutoModelForCausalLM, SepCache
        >>> import torch
        >>> from huggingface_hub import login
        >>> login("hf_xxxXXXxxx")


        >>> def to_cuda(a_dict: dict) -> dict:
        >>>    new_dict = {}    
        >>>    for k,v in a_dict.items():
        >>>        if isinstance(v, torch.Tensor):
        >>>            new_dict[k] = v.cuda()
        >>>        else:
        >>>            new_dict[k] = v
        >>>    return new_dict

        >>> model = AutoModelForCausalLM.from_pretrained("meta-llama/Meta-Llama-3-8B-Instruct", attn_implementation="flash_attention_2", device_map="cuda:0")
        >>> model.bfloat16().cuda()
        >>> tokenizer = AutoTokenizer.from_pretrained("meta-llama/Meta-Llama-3-8B-Instruct")
        >>> inputs = tokenizer(text="My name is Llama 3", return_tensors="pt")
        >>> inputs = to_cuda(inputs)
        >>> # Prepare a cache and pass it to model's forward; `layer_num` is the number of layers for the pretrained model.
        >>> past_key_values = SepCache(init_cache_size=4, sep_cache_size=64, local_size=256, cache_size=512, layer_num=32, USE_MAX_SEP_CACHE=True, model_type='llama')
        >>> # `separator_token_ids` and `PADDING_ID` must also be provided if you are not using `model_type='llama'` like this demo.
        >>> outputs = model(**inputs, past_key_values=past_key_values, use_cache=True)
        >>> outputs.past_key_values # access SepCache filled with keys/values
        SepCache()
        ```

        ```python
        >>> ## When using the `update` function of SepCache to update the keys/values and the past token ids (necessary in SepCache), the current `input_ids` must also be provided.        
        >>> key_states, value_states = past_key_values.update(                
                    key_states = key_states,
                    value_states = value_states,    
                    input_ids = input_ids,
                    layer_idx = layer_idx,     
                    PREFILLING_FLAG = q_len > 1, ## `q_len` is the sequence length of the current `query_states`
                    )

        ```
        For detailed usage instructions, please refer to https://github.com/HKUDS/SepLLM

 @ArthurZucker @gante 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by zucchini-nlp — COMMENTED** [review]
@GaussonTschen hey!

Super cool to see more cache methods, thanks for the PR.

Recently we added support for remote code in generation ([docs](https://huggingface.co/docs/transformers/generation_strategies#custom-decoding-methods)), which allows us to host any generation method or cache on the hub. It reduces the amount of long review cycles for you and you can share the new method right away from the hub. Users then can use the new cache by pointing to the hub repo

I think `SepCache` is another perfect candidate to test out the custom generation methods and we can add the method to `transformers` if it gets a lot of usage. See more under this comment https://github.com/huggingface/transformers/pull/38055#issuecomment-2900553588 and here is an example of [`SinkCache` repo](https://huggingface.co/transformers-community/sink_cache) on the hub.

LMK if you have any questions about the process :)

**Comment by GaussonTschen** [comment]
@zucchini-nlp Hello!
Thanks for your comments. I am happy to merge `SepCache` into `the custom generation methods`. However, there is something different about `SepCache` when compared to traditional `Cache` class like `SinkCache` you mentioned. The difference is as follows:

When using the `update` function of `SepCache` to update the keys/values and the past token ids (necessary in `SepCache`), the current `input_ids` must also be provided. 
```python
key_states, value_states, query_states = past_key_values.update(                
    key_states = key_states,
    value_states = value_states,    
    input_ids = input_ids,  # Required
    layer_idx = layer_idx,     
    PREFILLING_FLAG = q_len > 1, ## `q_len` is the sequence length of the current `query_states`
    cache_kwargs = None )
```

Therefore, it seems that it may not be as straightforward to customize the `model.generate()` function just by writing a "generate.py" file, like with `SinkCache`, because `SepCache` needs to interact with its model by passing `input_ids` to the `update()` function of `past_key_values`.  `update()` is typically called when computing Self-Attention in a class or function, implying that when a model opts to use `SepCache`, a `if-else branch` must be added in the `forward` function of the Attention calculation class in the model's `modeling.py` file to pass `input_ids` to the `update` function of `SepCache`.

Our intention is to merge `SepCache` into the main branch of `transformers` repo to offer it for use by other models or users, as its integration is expected to be straightforward and won't impact other parts of the project. We seek your advice on this matter.

Regarding the current pull request, we noticed that some automated tests did not pass (e.g., "2 workflows awaiting approval"). We have only added `SepCache` and its related declarations to `__init__.py` without modifying any other parts. Therefore, we would appreciate any advice you can provide, such as regarding code formatting, removing trailing spaces at the end of each line, etc., which we will review one by one.

Thank you very much  and look forward to your reply.

**Comment by zucchini-nlp** [comment]
cc @gante , in case we had similar community released cache classes hosted on the hub

**Comment by gante** [comment]
@GaussonTschen uhmmm... passing `input_ids` to `.update()` is a much bigger interface challenge -- we would need to add substantial model-level changes 😱 In our `model.forward` calls, we don't even guarantee the existence of `input_ids`. The attention layers, where the `.update()` method is called, also doesn't receive `input_ids` as an input.

I see two possible implementation paths:
1. Instead of a `custom_generate` method, this cache is demo'ed as a custom model. In the custom model repo, the model architecture is redefined to
    - require `input_ids` 
    - use the new cache by default
    - pipe `input_ids` all the way to the cache `.update` method
2. if the cache logic that uses `input_ids` can be used before/after the forward pass, then we can detach that part of the logic into a separate cache method, and call that method before/after the forward pass. This then can be shared with everyone through a `custom_generate` method (a custom generation loop that calls the new cache method in the right place)

**Comment by GaussonTschen** [comment]
> @GaussonTschen uhmmm... passing `input_ids` to `.update()` is a much bigger interface challenge -- we would need to add substantial model-level changes 😱 In our `model.forward` calls, we don't even guarantee the existence of `input_ids`. The attention layers, where the `.update()` method is called, also doesn't receive `input_ids` as an input.
> 
> I see two possible implementation paths:
> 
> 1. Instead of a `custom_generate` method, this cache is demo'ed as a custom model. In the custom model repo, the model architecture is redefined to
>    
>    * require `input_ids`
>    * use the new cache by default
>    * pipe `input_ids` all the way to the cache `.update` method
> 2. if the cache logic that uses `input_ids` can be used before/after the forward pass, then we can detach that part of the logic into a separate cache method, and call that method before/after the forward pass. This then can be shared with everyone through a `custom_generate` method (a custom generation loop that calls the new cache method in the right place)

@zucchini-nlp @gante 

Thank you very much for your response. I have implemented the `custom_generate` for `SepCache` and uploaded the code to https://huggingface.co/Gausson/sep_cache.

Question: How can I merge `https://huggingface.co/Gausson/sep_cache` into `transformers-community/sep_cache`? I couldn’t find a pull request option on the Hugging Face Hub—could you guide me on the correct procedure?

**Comment by GaussonTschen** [comment]
@zucchini-nlp Hello!
@gante

Thank you very much for your response. I have implemented the `custom_generate` for `SepCache` and uploaded the code to https://huggingface.co/Gausson/sep_cache.

Question: How can I merge `https://huggingface.co/Gausson/sep_cache` into `transformers-community/sep_cache`? I couldn’t find a pull request option on the Hugging Face Hub [`transformers-community`](https://huggingface.co/transformers-community)  — could you guide me on the correct procedure?

Besides, our method have got over 2000 downloads now😊



image

**Comment by zucchini-nlp** [comment]
Super coool, congrats on release! I'll let Joao help with that, not sure if we usually merge custom methods in `transformers-community` since it is a brand new feature

**Comment by gante** [comment]
@GaussonTschen centralizing all discussion for Sep Cache [here](https://huggingface.co/spaces/transformers-community/support/discussions/10) 🤗
